src/dune_rules/foreign.mli

```diff
@@ -182,3 +182,15 @@ module Sources : sig
       -> t
   end
 end
+
+module Object : sig
+  type t
+
+  val decode : t Dune_lang.Decoder.t
+end
+
+module Objects : sig
+  type t = (Loc.t * Object.t) list
```

Comment:
How about using a map? I think that's what we use in similar fields in the same source file.

Reply:
I'm not sure what the benefit of using  a map here would be, as we never index into this sequence. For clarity though, I've removed the `Objects` module and exposed `Object.build_path`.

Reply:
Yes, but it does give us the guarantee that there are no duplicate objects. Now that I recall this property, do we make sure that `(foreign_objects foo foo)` is rejected?

Reply:
Ah no we don't check for duplicate objects. Will add.

Reply:
Is there ever a case where the order of object files matters, and we'd want to preserve the order they're specified in the `dune` file when passing them on to the linker?

Reply:
I've updated this PR to use a map and raise an error if there are duplicate object names

Reply:
> Is there ever a case where the order of object files matters, and we'd want to preserve the order they're specified in the dune file when passing them on to the linker?

Good point. I think preserving the order given to us by the user is indeed the right behavior. So I guess a list is right after all?

Now I'm wondering if there are any other cases where we take liberties with linking order where we shouldn't.

Reply:
I changed the representation of `Foreign.Objects.t` back to a list and just added a check that it contains no duplicate names when it's constructed.